### ISSUE TITLE
Support diskless directly with bootc

### ISSUE THREAD

**Issue by cgwalters** [containers/bootc#898]
Chatting with someone here at KubeCon, we realized a nice way to diskless would be to use PXE boot, have bootc in the initramfs, then fetch the image to RAM and pivot into it.

This would replace squashfs with OCI which would be pretty cool.

**Comment by kfox1111** [comment]
This would be great. :)

**Comment by kfox1111** [comment]
It could potentially be used as a base implementation for https://github.com/kubevirt/kubevirt/issues/13212.

**Comment by scottcwang** [comment]
Dracut [`dmsquash-live`](https://github.com/dracutdevs/dracut/blob/master/modules.d/90dmsquash-live/parse-dmsquash-live.sh) and [`livenet`](https://github.com/dracutdevs/dracut/blob/master/modules.d/90livenet/parse-livenet.sh) currently support `root=live:LABEL=*`, `root=live:UUID=*`, `root=live:PARTLABEL=*`, `root=live:PARTUUID=*`, `root=live:CDLABEL=*`, `root=live:nfs*`, `root=live:http://*`, `root=live:https://*`, `root=live:ftp://*`, `root=live:torrent://*`, `root=live:tftp://*`. Perhaps some of this existing logic could get reused to parse the `root` dracut argument, find the container image, and pull it into the same RAM disk that dracut already uses for live booting.

**Comment by chaserhkj** [comment]
This issue looks stale but I think it is still cool. I do have a coreos diskless workstation deployment myself in my home lab so this is very relevant to me, too.  I'll see if I can make it work and make a contribution.

**Comment by chaserhkj** [comment]
Ok I got a proof-of-concept working [here](https://github.com/chaserhkj/bootc-live)

I wanted to make fedora coreos image to work with this first, but apparently when generating initramfs for live boot osbuild [did something extra](https://github.com/osbuild/osbuild/blob/f295fc5489eb47d22f97408c33e5a4e7c11e1cd1/stages/org.osbuild.coreos.live-artifacts.mono#L927) on top of the dracut-produced initramfs file and coreos dracut module is [depending on](https://github.com/coreos/fedora-coreos-config/blob/testing-devel/overlay.d/05core/usr/lib/dracut/modules.d/35coreos-live/is-live-image.sh) these extended stuff to pull the hooks and services correctly and setup for a live system.

I am thinking about making the stamp file myself and see what part of coreos-live machinery would just work with this (apparently there would be parts that won't work like the image mounting part). But maybe trying this with a simpler bootc image actually makes more sense.

**Comment by chaserhkj** [comment]
I switched the base image to test from fedora-coreos to fedora-bootc and it was much better, the project linked above should be usable now.

Just not very sure how this could be integrated into `bootc` itself and become a contribution.

Do we want to add extra dracut modules that ships with bootc and enables live boot? The trouble is that probably most of the bootc images need some distribution-specific modifications to work with live boot anyways.

Also since I am currently just use [umoci](https://github.com/opencontainers/umoci) for extracting the oci archive this is pulling some external dependency and I am not using any of the bootc tool code at all (after all it's really just reading a container image).

**Comment by cgwalters** [comment]
> Also since I am currently just use [umoci](https://github.com/opencontainers/umoci) for extracting the oci archive this is pulling some external dependency and I am not using any of the bootc tool code at all (after all it's really just reading a container image).

That's cool for a PoC work for sure, but I think we want `bootc status` and such to operate, and `/` should still be readonly by default, and ultimately we do want to support *soft* reboots into new updates - this type of stuff needs to go beyond just using umoci/skopeo to copy from the registry.

**Comment by cgwalters** [comment]
Thanks for working on this!! 

> Just not very sure how this could be integrated into bootc itself and become a contribution.
> Do we want to add extra dracut modules that ships with bootc and enables live boot? The trouble is that probably most of the bootc images need some distribution-specific modifications to work with live boot anyways.

We already do have a dracut module today for the initramfs setup for composefs. (Of course we shouldn't be tied to dracut, this can just be a systemd unit that runs in the initramfs, which dracut just installs, but so could another tool)

I think basically it'd accept e.g. `bootc.image.spec=registry.example.com/someimage@sha256:abcd123` or so as a karg.

WDYT?

**Comment by chaserhkj** [comment]
The POC is sound, I have used it for some while on my diskless workstation, but I totally agree that a proper integration into bootc project would need to do much more than the rough skopeo/umoci pull.

What is good is that I have been improving the POC for a bit as well so I feel some of the ideas and experiences there would be transferable. So what the POC is doing is that it tries to setup an ext4 partition on a zram device, pulling the image to it and unpacking the rootfs, then bind mount to make `/` read only and `/etc` `/var` read-write. This emulates a real bootc setup and somehow worked out well. Zram really helps a lot since the unpacking process might eat a lot of tmpfs space that could lead to a problem. We might want this too in the proper implementation.

In the POC I am also using kexec to load the kernel image from the pulled image as well. This eliminates the implicit requirement that the bootloader loaded kernel and the image kernel needs to be kmod-compatible. This might be the ultimate route bootc wants to pursue on the soft reboot direction, too.

For the proper implementation, I would envision we do it with composefs backend and extend the current initramfs module to:

1. Match a particular karg (`bootc.image.spec=`) to activate a new `setup-live-root` subcommand in the initramfs binary
2. Setup a zram device and create a fsverify-supporting FS on it, as physical root (maybe take another karg to determine which fs to use)
3. call `create_composefs_filesystem` and `commit_image` to pull the image to composefs repo on physical root and create rootfs image out of it
4. call `write_composefs_state` to write the deployment state in composefs repo
5. Extract the image id created in step 3 and hand over the normal booting process to `setup-root`.

If this sounds like a good plan, I could start with a draft PR on this.

**Comment by chaserhkj** [comment]
@cgwalters I might have enough bandwidth to start working on this this week, WDYT about the above plan?

**Comment by cgwalters** [comment]
Hi @chaserhkj glad to have you working on this! There's a few public meetings available as well as a Slack instance if you want to find others for real-time collaboration as well.

> Match a particular karg (bootc.image.spec=) 

Yes this needs some bike-shedding on naming/design. Close in the design space is https://www.freedesktop.org/software/systemd/man/latest/systemd-import-generator.html

> Setup a zram device and create a fsverify-supporting FS on it, as physical root (maybe take another karg to determine which fs to use)

Yeah, it's a bit annoying that tmpfs doesn't supports fsverity; I can't imagine it'd be good hard to add. Until that happens...we could target a loopback-mounted fs, it's the general default in this space anyways.

> call create_composefs_filesystem and commit_image to pull the image to composefs repo on physical root and create rootfs image out of it

I think this is basically `bootc install to-filesystem` from the initramfs` right?

> Extract the image id created in step 3 and hand over the normal booting process to setup-root.

I think we should be able to mount at `/sysroot` and switch root to it, the same as a regular boot.

**Comment by chaserhkj** [comment]
> I think this is basically bootc install to-filesystem from the initramfs` right?

Yes, but the problem with `install to-filesystem`'s current implementation is that the logic of bootloader installation is strongly coupled with the main procedure, but bootloader should be obviously a skip in live system. This might be a good time to do some refactor to separate that logic from the preparation of the filesystem itself. I have #1893 particularly tracking this issue as well. Other use cases like cross-arch installation might benefit a lot from this, too.

**Comment by alexlovelltroy** [comment]
Hi!  I'm part of the TSC for the [OpenCHAMI project](https://openchami.org).  Two members of the RedHat team spoke to our User Summit today and shared more about bootc.  They pointed me to this issue.  Very exciting project!

We're considering bootc for booting diskless HPC nodes.  Our current boot process involves iPXE -> https download of kernel and initrd -> nfsroot of multiple squashfs filesystems.  Our process to create the squashfs filesystems is buildah-based and resembles the bootc container build process.

However, when exploring booting containers directly, we hit a snag.  Transferring the whole container before we can boot is going to be a lot slower than our current boot process.  Even when we're looking at large squashfs filesystems (multiple 20-50G filesystems) the total transfer to get to multiuser with ssh is only ~400M because we only transfer the blocks we need.  As far as I know, there isn't an option in the OCI transport layer for block-level access.

This might be out of scope, but I'd be very interested in working with you to define OCI metadata/labels/layers that standardize how to find remote filesystems at boot time and a way to convey those from the container at boot time.  

We have also been thinking about putting the kernel and initrd in dedicated layers so we can pull them independently without pulling the rest of the container in the PXE phase.  Nothing formal to propose at the moment.  Wanted to share our thinking after the presentation we saw yesterday.

**Comment by chaserhkj** [comment]
@alexlovelltroy Thank you for sharing!

I have laid some foundational work on this feature last month but was taken by some other priorities. I'll see if I could find time to open a draft PR for this feature before this EOW.

For the transport layer you talked about, I think there are two mechanisms in OCI/bootc ecosystems that are close to what you are talking about: Rechunking and chunk-level de-duplication

So rechunking (done by [chunkah](https://github.com/coreos/chunkah)) is an approach to restructure container image layers in a way that related packages are grouped together in layers to increase the possibility of layer transfers hitting a local/registry cache. It requires insights into the packaging system of the image and afaik current tools are pretty limited to RPM systems only. There are a lot of efforts to make this general but I don't know the progress.

Chunk-level de-dup is likely closer to what you have with the squashfs images. This uses a rolling-window style algorithm that is closer to prior arts in rsync and papers like [FastCDC](https://www.usenix.org/conference/atc16/technical-sessions/presentation/xia). It is generally supported by the [podman ecosystem storage layer](https://github.com/containers/storage/blob/main/docs/containers-storage-zstd-chunked.md), but [not yet implemented](https://github.com/composefs/composefs-rs/pull/137) in bootc's storage backend.

The two systems are complementary [as pointed out here](https://github.com/coreos/chunkah?tab=readme-ov-file#relationship-to-zstdchunked), they deal with different levels of the data transfer.
